apps/dgst.c

```diff
@@ -468,7 +478,7 @@ static void show_digests(const OBJ_NAME *name, void *arg)
         return;
 
     /* Filter out message digests that we cannot use */
-    md = EVP_get_digestbyname(name->name);
+    md = EVP_MD_fetch(app_get0_libctx(), name->name, app_get0_propq());
```

Comment:
Not sure about this line working for an engine impl?

Reply:
Doesn't seem to affect it:

```
$ cat test/engine.cnf
openssl_conf = openssl_init

# Comment out the next line to ignore configuration errors
config_diagnostics = 1

[openssl_init]
providers = provider_sect
engines = engine_section

[provider_sect]
default = default_sect

[default_sect]
activate = 1

[engine_section]
pkcs11 = pkcs11_section

[pkcs11_section]
engine_id = pkcs11
dynamic_path = /tmp/engines-3.0/libpkcs11.so  #  in /tmp because this machine only has libpkcs11.so compiled against 1.1, so this one's copied from Fedora 36
MODULE_PATH = /usr/lib64/pkcs11/libsofthsm2.so
init = 0

$ (cd test && OPENSSL_CONF=engine.cnf ../util/wrap.pl ../apps/openssl dgst -engine pkcs11 -engine_impl -list)
Engine "pkcs11" set.
Supported digests:
-blake2b512                -blake2s256                -md5
-md5-sha1                  -sha1                      -sha224
-sha256                    -sha3-224                  -sha3-256
-sha3-384                  -sha3-512                  -sha384
-sha512                    -sha512-224                -sha512-256
-shake128                  -shake256                  -sm3
-ssl3-md5                  -ssl3-sha1
```

That's despite this engine not actually offering digests (or maybe it doesn't because I didn't configure a token):
```
$ (cd test && OPENSSL_CONF=engine.cnf ../util/wrap.pl ../apps/openssl dgst -engine pkcs11 -engine_impl -sha256 engine.cnf)
Engine "pkcs11" set.
Error setting digest
00A01FC5D97F0000:error:13000093:engine routines:ENGINE_get_digest:unimplemented digest:crypto/engine/tb_digest.c:78:
00A01FC5D97F0000:error:03000086:digital envelope routines:evp_md_init_internal:initialization error:crypto/evp/digest.c:322:
```

So in summary, this doesn't seem like a problem to me. When used with engines, it lists all available digests, when used with providers, it lists those that are available.